protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,241 @@
+no matter at which process.
+This was the main reason for introducing the determinism requirements on `ProcessProposal`
+in the ABCI 1.0 specification.
+
+If we are to relax the determinism requirements on the application,
+we first need to modify function $valid()$ to be of the form $valid(v, bc_{h-1}, x_p)$,
+where $x_p$ is a variable local to process $p$.
+As $x_p$ may be different from $x_q$ for two processes $p$ and $q$, using
+$valid(v, bc_{h-1}, x_p)$ to implement in `PrepareProposal`  and `ProcessProposal`
+may break Requirements 3 to 5.
+
+Consensus _validity_ property is then modified as follows:
+
+- _weak validity_: function $valid(v, bc_{h-1}, x_p)$ has returned _true_ at least once
+  by a correct process for the decided block $v$.
+
+### Eventual Requirements
+
+We now relax the relevant ABCI 1.0 requirements in the following way.
+
+* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:
+  There exists a time $ts_{h}$ for every height $h$ such that,
+  for any two correct processes $p$ and $q$ and any round $r_p$ in height $h$ starting after $ts_{h}$,
+  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
+  $q$'s application returns _Accept_ in `ResponseProcessProposal`.
+
+* The determinism-related requirements, namely requirements 4 and 5, are removed.
+
+We call $ts_{h}$ the coherence-stabilization time for height $h$.
+
+If we think in terms of $valid(v, bc_{h-1}, x_p)$, notice that it is the application's responsibility
+to ensure 3b, that is, the application designers need to prove that the $x_p$ values at correct processes
+are evolving in a way that eventually `ResponseProcessProposal` returns _Accept_ at all correct processes
+that call `RequestPropocessPropsal` for height $h$ after $ts_{h}$.
+For instance, in Proposer-Based Timestamp, $x_p$ can be considered to be process $p$'s local clock,
+and having clocks synchronized is the mechanism ensuring eventual acceptance of a proposal.
+Finally, it is worth noting that _weak validity_ requires just one correct processes while requirement 3b
+refers to all correct processes.
+
+> [TODO: can we _tighten up_ eventual coherence?... i.e., make it even weaker?]
+>
+> [TODO: Since we have removed determinism, can byzantine proposers cause mayhem now?]
+
+### Modifications to the consensus Algorithm
+
+The Tendermint algorithm as described in the [arXiv paper][arxiv],
+and as implemented in CometBFT up to version `v0.38.0-rc3`,
+cannot guarantee consensus _termination_ for applications
+that just fulfill requirement 3b (eventual coherence), but do not fulfill requirements 3, 4, and 5.
+
+We need the following modifications (in terms of the algorithm as described in page 6 of the arXiv paper):
+
+- remove the evaluation of `valid(v)` in lines 29, 36 and 50 (i.e. replace `valid(v)` by `true`)
+- modify line 23 as follows
+
+> _\[Original\]_ &nbsp; 23: **if** $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$ **then**
```

Comment:
Ok, lets go through the update process here:

The original version is: $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$.

This is the same as: $(valid(v) \land lockedRound_p = −1) \lor (valid(v) \land \lor lockedValue_p = v)$.

The first change is that we don't need to check the validity of a locked value. From this we get: $(valid(v) \land lockedRound_p = −1) \lor lockedValue_p = v$.

The second change is to avoid checking validity for a valid value, so: $((validValue_p = v \lor valid(v)) \land lockedRound_p = −1) \lor lockedValue_p = v$.

An alternative is to replace $valid(v)$ by a new version, say $valid*(v)$, so that we get back to a simpler representation: $(valid*(v) \land lockedRound_p = −1) \lor lockedValue_p = v$, where we define $valid*(v) = validValue = v \lor valid(v)$.

Reply:
Post-comment: the formulation on the document is **right**. This is a proposal for making it simpler.